src/Containers/container.jl

```diff
@@ -113,13 +113,51 @@ function container(f::Function, indices, ::Type{SparseAxisArray})
     mappings = Base.Generator(I -> I => f(I...), indices)
     # Same as `Dict(mapping)` but it will error if two indices are the same.
     data = NoDuplicateDict(mappings)
-    dict = data.dict
-    if isempty(dict)
-        # If `dict` is empty, it was not able to determine the type
-        # of the key hence the type of `dict` is `Dict{Any, Any}`.
-        # This is an issue since `SparseAxisArray` needs the key
-        # type to be a tuple.
-        dict = Dict{_eltype_or_any(indices),Any}()
-    end
-    return SparseAxisArray(dict)
+    return _sparseaxisarray(data.dict, f, indices)
+end
+
+# The NoDuplicateDict was able to infer the element type.
+_sparseaxisarray(dict::Dict, ::Any, ::Any) = SparseAxisArray(dict)
+
+function _container_dict(
+    K::Type{<:NTuple{N,Any}},
+    f::Function,
+    ::Type{<:NTuple{N,Any}},
+) where {N}
+    ret = Base.return_types(f, K)
```

Comment:
Could we get rid of it if we do `Base.@default_eltype(Base.Generator(f, indices))` ?

Reply:
At this point, we have the function `f` and we have argument type `K`, so there is no need for `@default_eltype`.  Calling `Base.@default_eltype(Base.Generator(f, indices))` earlier on doesn't work because it often returns something less-specific than the current approach.